// rdf-parse/lib/RdfParser.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RdfParser = void 0;
const core_1 = require("@comunica/core");
const stream_1 = require("stream");
/**
 * An RdfParser can parse any RDF serialization, based on a given content type.
 */
class RdfParser {
    constructor(args) {
        this.mediatorRdfParseMediatypes = args.mediatorRdfParseMediatypes;
        this.mediatorRdfParseHandle = args.mediatorRdfParseHandle;
    }
    /**
     * Get an array of all available content types for this parser.
     * @return {Promise<string[]>} A promise resolving to a string array of all content types.
     */
    getContentTypes() {
        return __awaiter(this, void 0, void 0, function* () {
            return Object.keys(yield this.getContentTypesPrioritized());
        });
    }
    /**
     * Get a hash of all available content types for this parser, mapped to a numerical priority.
     * @return {Promise<{[p: string]: number}>} A promise resolving to a hash mapping content type to a priority number.
     */
    getContentTypesPrioritized() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.mediatorRdfParseMediatypes.mediate({ context: core_1.ActionContext({}), mediaTypes: true })).mediaTypes;
        });
    }
    /**
     * Parse the given stream.
     * @param {NodeJS.ReadableStream} stream A string stream.
     * @param {IParseOptions} options Parsing options.
     * @return {Stream} An RDFJS quad stream.
     */
    parse(stream, options) {
        let contentType;
        if ('contentType' in options && options.contentType) {
            contentType = options.contentType;
        }
        else if ('path' in options && options.path) {
            contentType = this.getContentTypeFromExtension(options.path);
            if (!contentType) {
                throw new Error(`No valid extension could be detected from the given 'path' option: '${options.path}'`);
            }
        }
        else {
            throw new Error(`Missing 'contentType' or 'path' option while parsing.`);
        }
        // Create a new readable
        const readable = new stream_1.PassThrough({ objectMode: true });
        // Delegate parsing to the mediator
        this.mediatorRdfParseHandle.mediate({
            context: core_1.ActionContext(options),
            handle: { input: stream, baseIRI: options.baseIRI },
            handleMediaType: contentType,
        })
            .then((output) => {
            const quads = output.handle.quads;
            quads.on('error', (e) => readable.emit('error', e));
            quads.pipe(readable);
        })
            .catch((e) => readable.emit('error', e));
        return readable;
    }
    /**
     * Get the content type based on the extension of the given path,
     * which can be an URL or file path.
     * @param {string} path A path.
     * @return {string} A content type or the empty string.
     */
    getContentTypeFromExtension(path) {
        const dotIndex = path.lastIndexOf('.');
        if (dotIndex >= 0) {
            const ext = path.substr(dotIndex);
            // ignore dot
            return RdfParser.CONTENT_MAPPINGS[ext.substring(1)] || '';
        }
        return '';
    }
}
exports.RdfParser = RdfParser;
// tslint:disable:object-literal-sort-keys
RdfParser.CONTENT_MAPPINGS = {
    ttl: "text/turtle",
    turtle: "text/turtle",
    nt: "application/n-triples",
    ntriples: "application/n-triples",
    nq: "application/n-quads",
    nquads: "application/n-quads",
    rdf: "application/rdf+xml",
    rdfxml: "application/rdf+xml",
    owl: "application/rdf+xml",
    n3: "text/n3",
    trig: "application/trig",
    jsonld: "application/ld+json",
    json: "application/ld+json",
};
