// sparqlalgebrajs/lib/algebra.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expressionTypes = exports.types = void 0;
// TODO: add aggregates?
exports.types = {
    ALT: 'alt',
    ASK: 'ask',
    BGP: 'bgp',
    CONSTRUCT: 'construct',
    DESC: 'desc',
    DESCRIBE: 'describe',
    DISTINCT: 'distinct',
    EXPRESSION: 'expression',
    EXTEND: 'extend',
    FILTER: 'filter',
    FROM: 'from',
    GRAPH: 'graph',
    GROUP: 'group',
    INV: 'inv',
    JOIN: 'join',
    LEFT_JOIN: 'leftjoin',
    LINK: 'link',
    MINUS: 'minus',
    NOP: 'nop',
    NPS: 'nps',
    ONE_OR_MORE_PATH: 'OneOrMorePath',
    ORDER_BY: 'orderby',
    PATH: 'path',
    PATTERN: 'pattern',
    PROJECT: 'project',
    REDUCED: 'reduced',
    SEQ: 'seq',
    SERVICE: 'service',
    SLICE: 'slice',
    UNION: 'union',
    VALUES: 'values',
    ZERO_OR_MORE_PATH: 'ZeroOrMorePath',
    ZERO_OR_ONE_PATH: 'ZeroOrOnePath',
    COMPOSITE_UPDATE: 'compositeupdate',
    DELETE_INSERT: 'deleteinsert',
    LOAD: 'load',
    CLEAR: 'clear',
    CREATE: 'create',
    DROP: 'drop',
    ADD: 'add',
    MOVE: 'move',
    COPY: 'copy',
};
exports.expressionTypes = {
    AGGREGATE: 'aggregate',
    EXISTENCE: 'existence',
    NAMED: 'named',
    OPERATOR: 'operator',
    TERM: 'term',
    WILDCARD: 'wildcard',
};
