// @multiformats/blake2/cjs/blake2s.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var blakejs = require('blakejs');
var hasher = require('multiformats/hashes/hasher');
var multiformats = require('multiformats');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var blakejs__default = /*#__PURE__*/_interopDefaultLegacy(blakejs);

const {blake2s} = blakejs__default["default"];
const blake2s8 = hasher.from({
  name: 'blake2s-8',
  code: 45633,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 1))
});
const blake2s16 = hasher.from({
  name: 'blake2s-16',
  code: 45634,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 2))
});
const blake2s24 = hasher.from({
  name: 'blake2s-24',
  code: 45635,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 3))
});
const blake2s32 = hasher.from({
  name: 'blake2s-32',
  code: 45636,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 4))
});
const blake2s40 = hasher.from({
  name: 'blake2s-40',
  code: 45637,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 5))
});
const blake2s48 = hasher.from({
  name: 'blake2s-48',
  code: 45638,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 6))
});
const blake2s56 = hasher.from({
  name: 'blake2s-56',
  code: 45639,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 7))
});
const blake2s64 = hasher.from({
  name: 'blake2s-64',
  code: 45640,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 8))
});
const blake2s72 = hasher.from({
  name: 'blake2s-72',
  code: 45641,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 9))
});
const blake2s80 = hasher.from({
  name: 'blake2s-80',
  code: 45642,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 10))
});
const blake2s88 = hasher.from({
  name: 'blake2s-88',
  code: 45643,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 11))
});
const blake2s96 = hasher.from({
  name: 'blake2s-96',
  code: 45644,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 12))
});
const blake2s104 = hasher.from({
  name: 'blake2s-104',
  code: 45645,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 13))
});
const blake2s112 = hasher.from({
  name: 'blake2s-112',
  code: 45646,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 14))
});
const blake2s120 = hasher.from({
  name: 'blake2s-120',
  code: 45647,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 15))
});
const blake2s128 = hasher.from({
  name: 'blake2s-128',
  code: 45648,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 16))
});
const blake2s136 = hasher.from({
  name: 'blake2s-136',
  code: 45649,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 17))
});
const blake2s144 = hasher.from({
  name: 'blake2s-144',
  code: 45650,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 18))
});
const blake2s152 = hasher.from({
  name: 'blake2s-152',
  code: 45651,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 19))
});
const blake2s160 = hasher.from({
  name: 'blake2s-160',
  code: 45652,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 20))
});
const blake2s168 = hasher.from({
  name: 'blake2s-168',
  code: 45653,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 21))
});
const blake2s176 = hasher.from({
  name: 'blake2s-176',
  code: 45654,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 22))
});
const blake2s184 = hasher.from({
  name: 'blake2s-184',
  code: 45655,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 23))
});
const blake2s192 = hasher.from({
  name: 'blake2s-192',
  code: 45656,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 24))
});
const blake2s200 = hasher.from({
  name: 'blake2s-200',
  code: 45657,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 25))
});
const blake2s208 = hasher.from({
  name: 'blake2s-208',
  code: 45658,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 26))
});
const blake2s216 = hasher.from({
  name: 'blake2s-216',
  code: 45659,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 27))
});
const blake2s224 = hasher.from({
  name: 'blake2s-224',
  code: 45660,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 28))
});
const blake2s232 = hasher.from({
  name: 'blake2s-232',
  code: 45661,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 29))
});
const blake2s240 = hasher.from({
  name: 'blake2s-240',
  code: 45662,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 30))
});
const blake2s248 = hasher.from({
  name: 'blake2s-248',
  code: 45663,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 31))
});
const blake2s256 = hasher.from({
  name: 'blake2s-256',
  code: 45664,
  encode: input => multiformats.bytes.coerce(blake2s(input, undefined, 32))
});

exports.blake2s104 = blake2s104;
exports.blake2s112 = blake2s112;
exports.blake2s120 = blake2s120;
exports.blake2s128 = blake2s128;
exports.blake2s136 = blake2s136;
exports.blake2s144 = blake2s144;
exports.blake2s152 = blake2s152;
exports.blake2s16 = blake2s16;
exports.blake2s160 = blake2s160;
exports.blake2s168 = blake2s168;
exports.blake2s176 = blake2s176;
exports.blake2s184 = blake2s184;
exports.blake2s192 = blake2s192;
exports.blake2s200 = blake2s200;
exports.blake2s208 = blake2s208;
exports.blake2s216 = blake2s216;
exports.blake2s224 = blake2s224;
exports.blake2s232 = blake2s232;
exports.blake2s24 = blake2s24;
exports.blake2s240 = blake2s240;
exports.blake2s248 = blake2s248;
exports.blake2s256 = blake2s256;
exports.blake2s32 = blake2s32;
exports.blake2s40 = blake2s40;
exports.blake2s48 = blake2s48;
exports.blake2s56 = blake2s56;
exports.blake2s64 = blake2s64;
exports.blake2s72 = blake2s72;
exports.blake2s8 = blake2s8;
exports.blake2s80 = blake2s80;
exports.blake2s88 = blake2s88;
exports.blake2s96 = blake2s96;
