// @multiformats/blake2/cjs/blake2b.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var blakejs = require('blakejs');
var hasher = require('multiformats/hashes/hasher');
var multiformats = require('multiformats');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var blakejs__default = /*#__PURE__*/_interopDefaultLegacy(blakejs);

const {blake2b} = blakejs__default["default"];
const blake2b8 = hasher.from({
  name: 'blake2b-8',
  code: 45569,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 1))
});
const blake2b16 = hasher.from({
  name: 'blake2b-16',
  code: 45570,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 2))
});
const blake2b24 = hasher.from({
  name: 'blake2b-24',
  code: 45571,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 3))
});
const blake2b32 = hasher.from({
  name: 'blake2b-32',
  code: 45572,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 4))
});
const blake2b40 = hasher.from({
  name: 'blake2b-40',
  code: 45573,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 5))
});
const blake2b48 = hasher.from({
  name: 'blake2b-48',
  code: 45574,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 6))
});
const blake2b56 = hasher.from({
  name: 'blake2b-56',
  code: 45575,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 7))
});
const blake2b64 = hasher.from({
  name: 'blake2b-64',
  code: 45576,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 8))
});
const blake2b72 = hasher.from({
  name: 'blake2b-72',
  code: 45577,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 9))
});
const blake2b80 = hasher.from({
  name: 'blake2b-80',
  code: 45578,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 10))
});
const blake2b88 = hasher.from({
  name: 'blake2b-88',
  code: 45579,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 11))
});
const blake2b96 = hasher.from({
  name: 'blake2b-96',
  code: 45580,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 12))
});
const blake2b104 = hasher.from({
  name: 'blake2b-104',
  code: 45581,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 13))
});
const blake2b112 = hasher.from({
  name: 'blake2b-112',
  code: 45582,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 14))
});
const blake2b120 = hasher.from({
  name: 'blake2b-120',
  code: 45583,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 15))
});
const blake2b128 = hasher.from({
  name: 'blake2b-128',
  code: 45584,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 16))
});
const blake2b136 = hasher.from({
  name: 'blake2b-136',
  code: 45585,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 17))
});
const blake2b144 = hasher.from({
  name: 'blake2b-144',
  code: 45586,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 18))
});
const blake2b152 = hasher.from({
  name: 'blake2b-152',
  code: 45587,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 19))
});
const blake2b160 = hasher.from({
  name: 'blake2b-160',
  code: 45588,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 20))
});
const blake2b168 = hasher.from({
  name: 'blake2b-168',
  code: 45589,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 21))
});
const blake2b176 = hasher.from({
  name: 'blake2b-176',
  code: 45590,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 22))
});
const blake2b184 = hasher.from({
  name: 'blake2b-184',
  code: 45591,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 23))
});
const blake2b192 = hasher.from({
  name: 'blake2b-192',
  code: 45592,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 24))
});
const blake2b200 = hasher.from({
  name: 'blake2b-200',
  code: 45593,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 25))
});
const blake2b208 = hasher.from({
  name: 'blake2b-208',
  code: 45594,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 26))
});
const blake2b216 = hasher.from({
  name: 'blake2b-216',
  code: 45595,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 27))
});
const blake2b224 = hasher.from({
  name: 'blake2b-224',
  code: 45596,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 28))
});
const blake2b232 = hasher.from({
  name: 'blake2b-232',
  code: 45597,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 29))
});
const blake2b240 = hasher.from({
  name: 'blake2b-240',
  code: 45598,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 30))
});
const blake2b248 = hasher.from({
  name: 'blake2b-248',
  code: 45599,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 31))
});
const blake2b256 = hasher.from({
  name: 'blake2b-256',
  code: 45600,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 32))
});
const blake2b264 = hasher.from({
  name: 'blake2b-264',
  code: 45601,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 33))
});
const blake2b272 = hasher.from({
  name: 'blake2b-272',
  code: 45602,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 34))
});
const blake2b280 = hasher.from({
  name: 'blake2b-280',
  code: 45603,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 35))
});
const blake2b288 = hasher.from({
  name: 'blake2b-288',
  code: 45604,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 36))
});
const blake2b296 = hasher.from({
  name: 'blake2b-296',
  code: 45605,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 37))
});
const blake2b304 = hasher.from({
  name: 'blake2b-304',
  code: 45606,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 38))
});
const blake2b312 = hasher.from({
  name: 'blake2b-312',
  code: 45607,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 39))
});
const blake2b320 = hasher.from({
  name: 'blake2b-320',
  code: 45608,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 40))
});
const blake2b328 = hasher.from({
  name: 'blake2b-328',
  code: 45609,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 41))
});
const blake2b336 = hasher.from({
  name: 'blake2b-336',
  code: 45610,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 42))
});
const blake2b344 = hasher.from({
  name: 'blake2b-344',
  code: 45611,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 43))
});
const blake2b352 = hasher.from({
  name: 'blake2b-352',
  code: 45612,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 44))
});
const blake2b360 = hasher.from({
  name: 'blake2b-360',
  code: 45613,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 45))
});
const blake2b368 = hasher.from({
  name: 'blake2b-368',
  code: 45614,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 46))
});
const blake2b376 = hasher.from({
  name: 'blake2b-376',
  code: 45615,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 47))
});
const blake2b384 = hasher.from({
  name: 'blake2b-384',
  code: 45616,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 48))
});
const blake2b392 = hasher.from({
  name: 'blake2b-392',
  code: 45617,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 49))
});
const blake2b400 = hasher.from({
  name: 'blake2b-400',
  code: 45618,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 50))
});
const blake2b408 = hasher.from({
  name: 'blake2b-408',
  code: 45619,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 51))
});
const blake2b416 = hasher.from({
  name: 'blake2b-416',
  code: 45620,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 52))
});
const blake2b424 = hasher.from({
  name: 'blake2b-424',
  code: 45621,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 53))
});
const blake2b432 = hasher.from({
  name: 'blake2b-432',
  code: 45622,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 54))
});
const blake2b440 = hasher.from({
  name: 'blake2b-440',
  code: 45623,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 55))
});
const blake2b448 = hasher.from({
  name: 'blake2b-448',
  code: 45624,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 56))
});
const blake2b456 = hasher.from({
  name: 'blake2b-456',
  code: 45625,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 57))
});
const blake2b464 = hasher.from({
  name: 'blake2b-464',
  code: 45626,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 58))
});
const blake2b472 = hasher.from({
  name: 'blake2b-472',
  code: 45627,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 59))
});
const blake2b480 = hasher.from({
  name: 'blake2b-480',
  code: 45628,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 60))
});
const blake2b488 = hasher.from({
  name: 'blake2b-488',
  code: 45629,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 61))
});
const blake2b496 = hasher.from({
  name: 'blake2b-496',
  code: 45630,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 62))
});
const blake2b504 = hasher.from({
  name: 'blake2b-504',
  code: 45631,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 63))
});
const blake2b512 = hasher.from({
  name: 'blake2b-512',
  code: 45632,
  encode: input => multiformats.bytes.coerce(blake2b(input, undefined, 64))
});

exports.blake2b104 = blake2b104;
exports.blake2b112 = blake2b112;
exports.blake2b120 = blake2b120;
exports.blake2b128 = blake2b128;
exports.blake2b136 = blake2b136;
exports.blake2b144 = blake2b144;
exports.blake2b152 = blake2b152;
exports.blake2b16 = blake2b16;
exports.blake2b160 = blake2b160;
exports.blake2b168 = blake2b168;
exports.blake2b176 = blake2b176;
exports.blake2b184 = blake2b184;
exports.blake2b192 = blake2b192;
exports.blake2b200 = blake2b200;
exports.blake2b208 = blake2b208;
exports.blake2b216 = blake2b216;
exports.blake2b224 = blake2b224;
exports.blake2b232 = blake2b232;
exports.blake2b24 = blake2b24;
exports.blake2b240 = blake2b240;
exports.blake2b248 = blake2b248;
exports.blake2b256 = blake2b256;
exports.blake2b264 = blake2b264;
exports.blake2b272 = blake2b272;
exports.blake2b280 = blake2b280;
exports.blake2b288 = blake2b288;
exports.blake2b296 = blake2b296;
exports.blake2b304 = blake2b304;
exports.blake2b312 = blake2b312;
exports.blake2b32 = blake2b32;
exports.blake2b320 = blake2b320;
exports.blake2b328 = blake2b328;
exports.blake2b336 = blake2b336;
exports.blake2b344 = blake2b344;
exports.blake2b352 = blake2b352;
exports.blake2b360 = blake2b360;
exports.blake2b368 = blake2b368;
exports.blake2b376 = blake2b376;
exports.blake2b384 = blake2b384;
exports.blake2b392 = blake2b392;
exports.blake2b40 = blake2b40;
exports.blake2b400 = blake2b400;
exports.blake2b408 = blake2b408;
exports.blake2b416 = blake2b416;
exports.blake2b424 = blake2b424;
exports.blake2b432 = blake2b432;
exports.blake2b440 = blake2b440;
exports.blake2b448 = blake2b448;
exports.blake2b456 = blake2b456;
exports.blake2b464 = blake2b464;
exports.blake2b472 = blake2b472;
exports.blake2b48 = blake2b48;
exports.blake2b480 = blake2b480;
exports.blake2b488 = blake2b488;
exports.blake2b496 = blake2b496;
exports.blake2b504 = blake2b504;
exports.blake2b512 = blake2b512;
exports.blake2b56 = blake2b56;
exports.blake2b64 = blake2b64;
exports.blake2b72 = blake2b72;
exports.blake2b8 = blake2b8;
exports.blake2b80 = blake2b80;
exports.blake2b88 = blake2b88;
exports.blake2b96 = blake2b96;
