// quadstore/dist/lib/serialization/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.copyBuffer = exports.sliceBuffer = exports.copyBufferIntoBuffer = exports.writePattern = exports.quadReader = exports.quadWriter = void 0;
var quads_1 = require("./quads");
Object.defineProperty(exports, "quadWriter", { enumerable: true, get: function () { return quads_1.quadWriter; } });
Object.defineProperty(exports, "quadReader", { enumerable: true, get: function () { return quads_1.quadReader; } });
var patterns_1 = require("./patterns");
Object.defineProperty(exports, "writePattern", { enumerable: true, get: function () { return patterns_1.writePattern; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "copyBufferIntoBuffer", { enumerable: true, get: function () { return utils_1.copyBufferIntoBuffer; } });
Object.defineProperty(exports, "sliceBuffer", { enumerable: true, get: function () { return utils_1.sliceBuffer; } });
Object.defineProperty(exports, "copyBuffer", { enumerable: true, get: function () { return utils_1.copyBuffer; } });
