// quadstore/dist/lib/serialization/xsd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.positiveInteger = exports.unsignedByte = exports.unsignedShort = exports.unsignedInt = exports.unsignedLong = exports.nonNegativeInteger = exports.byte = exports.short = exports.int = exports.long = exports.negativeInteger = exports.nonPositiveInteger = exports.double = exports.decimal = exports.integer = exports.boolean = exports.dateTime = exports.string = exports.langString = exports.rdf = exports.xsd = void 0;
exports.xsd = 'http://www.w3.org/2001/XMLSchema#';
exports.rdf = 'http://www.w3.org/1999/02/22-rdf-syntax-ns#';
exports.langString = `${exports.rdf}langString`;
exports.string = `${exports.xsd}string`;
exports.dateTime = `${exports.xsd}dateTime`;
exports.boolean = `${exports.xsd}boolean`;
exports.integer = `${exports.xsd}integer`;
exports.decimal = `${exports.xsd}decimal`;
exports.double = `${exports.xsd}double`;
exports.nonPositiveInteger = `${exports.xsd}nonPositiveInteger`;
exports.negativeInteger = `${exports.xsd}negativeInteger`;
exports.long = `${exports.xsd}long`;
exports.int = `${exports.xsd}int`;
exports.short = `${exports.xsd}short`;
exports.byte = `${exports.xsd}byte`;
exports.nonNegativeInteger = `${exports.xsd}nonNegativeInteger`;
exports.unsignedLong = `${exports.xsd}unsignedLong`;
exports.unsignedInt = `${exports.xsd}unsignedInt`;
exports.unsignedShort = `${exports.xsd}unsignedShort`;
exports.unsignedByte = `${exports.xsd}unsignedByte`;
exports.positiveInteger = `${exports.xsd}positiveInteger`;
