// @comunica/core/lib/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KEY_CONTEXT_LOG = exports.Logger = void 0;
const context_entries_1 = require("@comunica/context-entries");
/**
 * A logger accepts messages from different levels
 * and emits them in a certain way.
 */
class Logger {
    /**
     * Convert a string-based logging level to a numerical logging level.
     * @param level A string-based logging level
     * @return The numerical logging level, or undefined.
     */
    static getLevelOrdinal(level) {
        return Logger.LEVELS[level];
    }
}
exports.Logger = Logger;
/**
 * All available logging levels.
 * @type {{trace: number; debug: number; info: number; warn: number; error: number; fatal: number}}
 */
Logger.LEVELS = {
    trace: 0,
    debug: 1,
    info: 2,
    warn: 3,
    error: 4,
    fatal: 5,
};
/**
 * @type {string} Context entry for a logger instance.
 * @value {Logger} A logger.
 * @deprecated Import this constant from @comunica/context-entries.
 */
exports.KEY_CONTEXT_LOG = context_entries_1.KeysCore.log;
