// @multiformats/sha3/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var hasher = require('multiformats/hashes/hasher');
var sha3$1 = require('js-sha3');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var sha3__default = /*#__PURE__*/_interopDefaultLegacy(sha3$1);

'use strict';
function encoder(fn) {
  return b => new Uint8Array(fn.array(b));
}
const sha3224 = hasher.from({
  code: 23,
  name: 'sha3-224',
  encode: encoder(sha3__default["default"].sha3_224)
});
const sha3256 = hasher.from({
  code: 22,
  name: 'sha3-256',
  encode: encoder(sha3__default["default"].sha3_256)
});
const sha3384 = hasher.from({
  code: 21,
  name: 'sha3-384',
  encode: encoder(sha3__default["default"].sha3_384)
});
const sha3512 = hasher.from({
  code: 20,
  name: 'sha3-512',
  encode: encoder(sha3__default["default"].sha3_512)
});
const shake128 = hasher.from({
  code: 24,
  name: 'shake-128',
  encode: b => new Uint8Array(sha3__default["default"].shake128.array(b, 256))
});
const shake256 = hasher.from({
  code: 25,
  name: 'shake-256',
  encode: b => new Uint8Array(sha3__default["default"].shake256.array(b, 512))
});
const keccak224 = hasher.from({
  code: 26,
  name: 'keccak-224',
  encode: encoder(sha3__default["default"].keccak224)
});
const keccak256 = hasher.from({
  code: 27,
  name: 'keccak-256',
  encode: encoder(sha3__default["default"].keccak256)
});
const keccak384 = hasher.from({
  code: 28,
  name: 'keccak-384',
  encode: encoder(sha3__default["default"].keccak384)
});
const keccak512 = hasher.from({
  code: 29,
  name: 'keccak-512',
  encode: encoder(sha3__default["default"].keccak512)
});

exports.keccak224 = keccak224;
exports.keccak256 = keccak256;
exports.keccak384 = keccak384;
exports.keccak512 = keccak512;
exports.sha3224 = sha3224;
exports.sha3256 = sha3256;
exports.sha3384 = sha3384;
exports.sha3512 = sha3512;
exports.shake128 = shake128;
exports.shake256 = shake256;
